File: koma-cli/src/main/java/org/koma/cli/lifecycle/LoggerLifecycle.java
package org.koma.cli.lifecycle;

import ch.qos.logback.classic.LoggerContext;
import com.google.common.util.concurrent.AbstractIdleService;
import jakarta.inject.Singleton;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.LoggerFactory;

@Singleton
@Slf4j
public class LoggerLifecycle extends AbstractIdleService {

  @Override
  protected void startUp() throws Exception {
    if (LoggerFactory.getILoggerFactory() instanceof LoggerContext context) {
      log.atDebug().log("LoggerLifecycle init");
    }
  }

  @Override
  protected void shutDown() throws Exception {
    log.atInfo().log("Pre-destroy");
    if (LoggerFactory.getILoggerFactory() instanceof LoggerContext context) {
      context.stop();
    }
  }
}

File: koma-cli/src/main/kotlin/org/koma/cli/lifecycle/LoggerLifecycle.kt
package org.koma.cli.lifecycle

import ch.qos.logback.classic.LoggerContext
import com.google.common.util.concurrent.AbstractIdleService
import io.github.oshai.kotlinlogging.KotlinLogging
import lombok.extern.slf4j.Slf4j
import org.koin.core.annotation.Single
import org.slf4j.LoggerFactory


@Single
class LoggerLifecycle : AbstractIdleService() {
  private val log = KotlinLogging.logger {}

  override fun startUp() {
    if (LoggerFactory.getILoggerFactory() is LoggerContext) {
      log.atDebug { message = "LoggerLifecycle init" }
    }
  }

  override fun shutDown() {
    val context = LoggerFactory.getILoggerFactory()
    if (context is LoggerContext) {
      context.stop()
    }
  }
}
